ext/intl/formatter/formatter_format.c

```diff
@@ -53,30 +53,55 @@ PHP_FUNCTION( numfmt_format )
 			case IS_DOUBLE:
 				type = FORMAT_TYPE_DOUBLE;
 				break;
-			EMPTY_SWITCH_DEFAULT_CASE();
+			case IS_STRING:
+				type = FORMAT_TYPE_DECIMAL;
+				break;
+			default:
+				zend_argument_type_error(1, "must be of type int|float|string, %s given", zend_zval_type_name(number));
 		}
 	}
 
+	// Avoid losing precision on 32-bit platforms where PHP's "long" isn't
+	// as long as the FORMAT_TYPE_INT64 which is requested.
+#if SIZEOF_ZEND_LONG < 8
+	if (Z_TYPE_P(number) == IS_STRING && type == FORMAT_TYPE_INT64) {
+		type = FORMAT_TYPE_DECIMAL;
+	}
+#endif
+
 	switch(type) {
 		case FORMAT_TYPE_INT32:
+		{
+			bool failed = true;
+			int64_t value_64 = zval_try_get_long(number, &failed);
+			if (failed || value_64 < -2147483648 || value_64 > 2147483647) {
+				zend_argument_value_error(object ? 1 : 2, "must be numeric and fit in 32 bits");
+				RETURN_THROWS();
+			}
```

Comment:
The ``failed`` case should be split into a different error, as it is a type error.